resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -144,15 +144,96 @@ public void operationComplete(Future<T> future) throws Exception {
     }
 
     private void internalResolve(Promise<T> promise) {
-        InetSocketAddress nameServerAddrToTry = nameServerAddrs.next();
+        DnsServerAddressStream nsAddrs = getNameServers(hostname);
+
         for (DnsRecordType type: parent.resolveRecordTypes()) {
-            if (!query(hostname, type, nameServerAddrToTry, promise)) {
+            if (!query(hostname, type, nsAddrs, promise)) {
                 return;
             }
         }
     }
 
-    private void query(InetSocketAddress nameServerAddr, final DnsQuestion question, final Promise<T> promise) {
+    /**
+     * Add an authoritative nameserver to the cache if its not a root server.
+     */
+    private void addNameServerToCache(
+            AuthoritativeNameServerList.AuthoritativeNameServer name, InetAddress resolved, long ttl) {
+        if (!name.isRootServer()) {
+            // Cache NS record if not for a root server as we should never cache for root servers.
+            parent.authoritativeDnsServerCache().cache(name.domainName(),
+                    additionals, resolved, ttl, parent.ch.eventLoop());
+        }
+    }
+
+    /**
+     * Returns the {@link DnsServerAddressStream} that was cached for the given hostname or {@code null} if non
+     *  could be found.
+     */
+    private DnsServerAddressStream getNameServersFromCache(String hostname) {
+        int len = hostname.length();
+
+        if (len == 0 || len == 1 && hostname.charAt(0) == '.') {
+            // We never cache for root servers.
+            return null;
+        }
+
+        // We always store in the case with a trailing '.'.
+        if (hostname.charAt(len - 1) != '.') {
+            hostname += ".";
```

Comment:
I think its easier here as otherwise we may need to do it multiple times when call `parent.authoritativeDnsServerCache().get(hostname, additionals);`